src/vm/amd64/umthunkstub.S

```diff
@@ -0,0 +1,551 @@
+
+// Now we have the full size of the stack frame.  The offsets have been computed relative to the
+// top, so negate them to make them relative to the post-prologue rsp.
+#define UMThunkStubAMD64_FRAME_OFFSET UMThunkStubAMD64_CALLEE_SCRATCH_SIZE
+#define UMThunkStubAMD64_RARE_PATH_SPILL_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_RARE_PATH_SPILL_NEGOFFSET)
+#define UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_HOST_NOTIFY_FLAG_NEGOFFSET)
+#define UMThunkStubAMD64_XMM_SAVE_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE - UMThunkStubAMD64_FRAME_OFFSET - UMThunkStubAMD64_XMM_SAVE_NEGOFFSET)
+#define UMThunkStubAMD64_ARGUMENTS_STACK_HOME_OFFSET (UMThunkStubAMD64_STACK_FRAME_SIZE + 8 - UMThunkStubAMD64_FRAME_OFFSET)
+#define UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE (UMThunkStubAMD64_STACK_FRAME_SIZE - (UMThunkStubAMD64_NUM_REG_PUSHES * 8))
+
+#if (UMTHUNKSTUB_HOST_NOTIFY_FLAG_RBPOFFSET - UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET) != 0
+#error update UMTHUNKSTUB_HOST_NOTIFY_FLAG_RBPOFFSET
+#endif
+
+//
+// [ callee scratch ]            <-- new RSP
+// [ callee scratch ]
+// [ callee scratch ]
+// [ callee scratch ]
+// {optional stack args passed to callee}
+// xmm0                          <-- RBP
+// xmm1
+// xmm2
+// xmm3
+// xmm4
+// xmm5
+// xmm6
+// xmm7
+// {optional padding to align xmm regs}
+// HOST_NOTIFY_FLAG (needs to make ReverseLeaveRuntime call flag)
+// [rare path spill area]
+// [rare path spill area]
+// rbp save
+// r12 save
+// return address                <-- entry RSP
+// [rdi home]
+// [rsi home]
+// [rdx home]
+// [rcx home]
+// [r8 home]
+// [r9 home]
+// stack arg 0
+// stack arg 1
+// ...
+
+        push_nonvol_reg r12
+        push_nonvol_reg rbp                                                                     // stack_args
+        alloc_stack     UMThunkStubAMD64_FIXED_STACK_ALLOC_SIZE
+        set_frame       rbp, UMThunkStubAMD64_FRAME_OFFSET                                      // stack_args
+        mov             byte ptr [rbp + UMThunkStubAMD64_HOST_NOTIFY_FLAG_OFFSET], 0            // hosted
+        END_PROLOGUE
+
+        //
+        // Call GetThread()
+        //
+        call            C_FUNC(GetThread)                          // will not trash r10
+        test            rax, rax
+        jz              LOCAL_LABEL(DoThreadSetup)
+
+LOCAL_LABEL(HaveThread):
+
+        mov             r12, rax                // r12 <- Thread*
+
+        //
+        // disable preemptive GC
+        //
+        mov             dword ptr [r12 + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 1
+
+        //
+        // catch returning thread here if a GC is in progress
+        //
+        cmp             [rip + C_FUNC(g_TrapReturningThreads)@GOTPCREL], 0
+        jnz             LOCAL_LABEL(DoTrapReturningThreadsTHROW)
+
+LOCAL_LABEL(InCooperativeMode):
+
+#ifdef FEATURE_INCLUDE_ALL_INTERFACES
```

Comment:
FEATURE_INCLUDE_ALL_INTERFACES is old feature of full .NET Framework that is unlikely to be ever needed in CoreCLR or cross-platform. This code can be deleted here.